### PR TITLE
Fix that leaves were being dequeued in reverse order. Ooops.

### PR DESCRIPTION
**Pull request by Martin2112**
And add a test that checks for the correct ordering.

### CODE REVIEW

File: storage/mysql/storage_test.go [hunk 1]
```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
...
+
+	logID := createLogID("TestDequeueLeavesTimeOrdering")
+	db := prepareTestLogDB(logID, t)
+	defer db.Close()
+	s := prepareTestLogStorage(logID, t)
+	batchSize := 2
+
+	{
+		tx := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx)
```
Comment (AlCutter) [inline storage/mysql/storage_test.go]: `defer`ed funcs only run at the return of the enclosing function, not when the enclosing scope goes away, does that change anything here?

File: storage/mysql/storage_test.go [hunk 2]
```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
...
+
+		leaves := createTestLeaves(int64(batchSize), 0)
+		leaves2 := createTestLeaves(int64(batchSize), int64(batchSize))
+
+		if err := tx.QueueLeaves(leaves, fakeQueueTime); err != nil {
+			t.Fatalf("Failed to queue leaves: %v", err)
```
Comment (AlCutter) [inline storage/mysql/storage_test.go]: `"QueueLeaves(..., ...) = %v", err`

File: storage/mysql/storage_test.go [hunk 3]
```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
...
+		}
+
+		// These are one second earlier so should be dequeued first
+		if err := tx.QueueLeaves(leaves2, fakeQueueTime.Add(-time.Second)); err != nil {
+			t.Fatalf("Failed to queue leaves2: %v", err)
```
Comment (AlCutter) [inline storage/mysql/storage_test.go]: `"QueueLeaves(..., ...) = %v", err`

File: storage/mysql/storage_test.go [hunk 4]
```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
...
+	}
+
+	{
+		// Now try to dequeue two leaves and we should get the second batch
+		tx2 := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx2)
+		dequeue1, err := tx2.DequeueLeaves(batchSize, fakeQueueTime)
+
+		if err != nil {
+			t.Fatalf("Failed to dequeue leaves (1st): %v", err)
```
Comment (AlCutter) [inline storage/mysql/storage_test.go]: error message style (and elsewhere)

File: storage/mysql/storage_test.go [hunk 5]
```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
...
+
+		if got, want := len(dequeue1), batchSize; got != want {
+			t.Fatalf("Dequeue count mismatch (1st) got: %d, want: %d", got, want)
+		}
+
+		ensureAllLeafHashesDistinct(dequeue1, t)
+
+		// Ensure this is the second batch queued by comparing leaf data.
+		if !leafInRange(dequeue1[0], batchSize, batchSize + batchSize - 1) || !leafInRange(dequeue1[1], batchSize, batchSize + batchSize - 1) {
+			t.Fatalf("Got leaf from wrong batch (1st dequeue): (%s %s) %v", string(dequeue1[0].LeafValue), string(dequeue1[1].LeafValue), dequeue1)
```
Comment (AlCutter) [inline storage/mysql/storage_test.go]: Possibly don't need that last `%v` and corresponding `dequeue1`? (or don't need the individual elements)

File: storage/mysql/storage_test.go (hunk 1 above)
Comment (Martin2112) [inline storage/mysql/storage_test.go]: No, they should all get committed by the end of the test. But I missed the defer for tx3 so I added that one.

File: storage/mysql/storage_test.go (hunk 2 above)
Comment (Martin2112) [inline storage/mysql/storage_test.go]: Done.

File: storage/mysql/storage_test.go (hunk 3 above)
Comment (Martin2112) [inline storage/mysql/storage_test.go]: Done.

File: storage/mysql/storage_test.go (hunk 4 above)
Comment (Martin2112) [inline storage/mysql/storage_test.go]: They're not error types though.

File: storage/mysql/storage_test.go (hunk 5 above)
Comment (Martin2112) [inline storage/mysql/storage_test.go]: It was originally just %v but the output was hard to read so I pulled out the mismatched data.

File: storage/mysql/storage_test.go (hunk 4 above)
Comment (AlCutter) [inline storage/mysql/storage_test.go]: Sorry, I meant:
   change to "DequeueLeaves(...) = %v ..." style messages (and elsewhere)
but I was too lazy to type it all out again :)

File: storage/mysql/storage_test.go (hunk 5 above)
Comment (AlCutter) [inline storage/mysql/storage_test.go]: `%+v` or `%#v` might be better?

Comment (Martin2112) [inline storage/mysql/storage_test.go]: I'll make it just print the mismatched strings. The rest of the leaf contents don't really matter in this context.

Comment (Martin2112) [inline storage/mysql/storage_test.go]: Done.

**Comment by Martin2112** [comment]
Is this OK yet?

**Comment by AlCutter** [comment]
Mostly, could you run gofmt on it and see what (if anything) it does on the `a+b` vs `a + b` front?

**Comment by Martin2112** [comment]
Ran gofmt again from command line and it did change a+b. Speculative theory is that the intellij go fmt is interacting with the "format changed lines only" editor setting.

**Pull request merged**